oap-server/server-storage-plugin/storage-iotdb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/iotdb/query/IoTDBTraceQueryDAO.java

```diff
@@ -0,0 +1,156 @@
+            query.append(" and ").append(SegmentRecord.ENDPOINT_ID).append(" = \"").append(endpointId).append("\"");
+        }
+        if (CollectionUtils.isNotEmpty(tags)) {
+            for (final Tag tag : tags) {
+                query.append(" and ").append(tag.getKey()).append(" = \"").append(tag.getValue()).append("\"");
+            }
+        }
+        switch (traceState) {
+            case ERROR:
+                query.append(" and ").append(SegmentRecord.IS_ERROR).append(" = ").append(BooleanUtils.TRUE);
+                break;
+            case SUCCESS:
+                query.append(" and ").append(SegmentRecord.IS_ERROR).append(" = ").append(BooleanUtils.FALSE);
+                break;
+        }
+        // IoTDB doesn't support the query contains "1=1" and "*" at the meantime.
+        String queryString = query.toString().replace("1=1 and ", "");
+        queryString = queryString + IoTDBClient.ALIGN_BY_DEVICE;
+
+        TraceBrief traceBrief = new TraceBrief();
+        List<? super StorageData> storageDataList = client.filterQuery(SegmentRecord.INDEX_NAME, queryString, storageBuilder);
+        int limitCount = 0;
+        for (int i = 0; i < storageDataList.size(); i++) {
+            if (i >= from && limitCount < limit) {
+                limitCount++;
+                SegmentRecord segmentRecord = (SegmentRecord) storageDataList.get(i);
+                BasicTrace basicTrace = new BasicTrace();
+                basicTrace.setSegmentId(segmentRecord.getSegmentId());
+                basicTrace.setStart(String.valueOf(segmentRecord.getStartTime()));
+                basicTrace.getEndpointNames().add(IDManager.EndpointID.analysisId(segmentRecord.getEndpointId()).getEndpointName());
+                basicTrace.setDuration(segmentRecord.getLatency());
+                basicTrace.setError(BooleanUtils.valueToBoolean(segmentRecord.getIsError()));
+                basicTrace.getTraceIds().add(segmentRecord.getTraceId());
+                traceBrief.getTraces().add(basicTrace);
+            }
+        }
+        traceBrief.setTotal(storageDataList.size());
+        // resort by self, because of the select query result order by time.
+        switch (queryOrder) {
+            case BY_START_TIME:
+                traceBrief.getTraces().sort((BasicTrace b1, BasicTrace b2) ->
+                        Long.compare(Long.parseLong(b2.getStart()), Long.parseLong(b1.getStart())));
+                break;
+            case BY_DURATION:
+                traceBrief.getTraces().sort((BasicTrace b1, BasicTrace b2) ->
+                        Integer.compare(b2.getDuration(), b1.getDuration()));
+                break;
+        }
+        return traceBrief;
+    }
+
+    @Override
+    public List<SegmentRecord> queryByTraceId(String traceId) throws IOException {
+        StringBuilder query = new StringBuilder();
+        query.append("select * from ").append(client.getStorageGroup()).append(IoTDBClient.DOT).append(SegmentRecord.INDEX_NAME);
```

Comment:
Consider adding an index formatter instead of concatenating this part over and over, like https://github.com/apache/skywalking/blob/4a344b01e9ad103b1cfda3e56d774dd4778c77de/oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/StorageModuleElasticsearchProvider.java#L225-L232